vpr/src/route/rr_graph_storage.h

```diff
@@ -0,0 +1,578 @@
+
+    /* Edge finalization methods */
+
+    // Counts the number of rr switches needed based on fan in.
+    //
+    // init_fan_in does not need to be invoked before this method.
+    size_t count_rr_switches(
+        size_t num_arch_switches,
+        t_arch_switch_inf* arch_switch_inf,
+        t_arch_switch_fanin& arch_switch_fanins);
+
+    // Maps arch_switch_inf indicies to rr_switch_inf indicies.
+    //
+    // This must be called before partition_edges if edges were created with
+    // arch_switch_inf indicies.
+    void remap_rr_node_switch_indices(const t_arch_switch_fanin& switch_fanin);
+
+    // Marks that edge switch values are rr switch indicies.
+    //
+    // This must be called before partition_edges if edges were created with
+    // rr_switch_inf indicies.
+    void mark_edges_as_rr_switch_ids();
+
+    // Sorts edge data such that configurable edges appears before
+    // non-configurable edges.
+    void partition_edges();
+
+    // Validate that edge data is partitioned correctly.
+    bool validate() const;
+
+    /* Edge accessors
+     *
+     * Preferred access methods:
+     * - first_edge(RRNodeId)
+     * - last_edge(RRNodeId)
+     * - edge_sink_node(RREdgeId)
+     * - edge_switch(RREdgeId)
+     *
+     * Legacy/deprecated access methods:
+     * - edges(RRNodeId)
+     * - configurable_edges(RRNodeId)
+     * - non_configurable_edges(RRNodeId)
+     * - num_edges(RRNodeId)
+     * - num_configurable_edges(RRNodeId)
+     * - num_non_configurable_edges(RRNodeId)
+     * - edge_id(RRNodeId, t_edge_size)
+     * - edge_sink_node(RRNodeId, t_edge_size)
+     * - edge_switch(RRNodeId, t_edge_size)
+     *
+     * Only call these methods after partition_edges has been invoked. */
+    edge_idx_range edges(const RRNodeId& id) const {
+        return vtr::make_range(edge_idx_iterator(0), edge_idx_iterator(num_edges(id)));
+    }
+    edge_idx_range configurable_edges(const RRNodeId& id) const {
+        return vtr::make_range(edge_idx_iterator(0), edge_idx_iterator(num_edges(id) - num_non_configurable_edges(id)));
+    }
+    edge_idx_range non_configurable_edges(const RRNodeId& id) const {
+        return vtr::make_range(edge_idx_iterator(num_edges(id) - num_non_configurable_edges(id)), edge_idx_iterator(num_edges(id)));
+    }
+
+    t_edge_size num_edges(const RRNodeId& id) const {
+        return size_t(last_edge(id)) - size_t(first_edge(id));
+    }
+    t_edge_size num_configurable_edges(const RRNodeId& id) const;
+    t_edge_size num_non_configurable_edges(const RRNodeId& id) const;
+
+    // Get the first and last RREdgeId for the specified RRNodeId.
+    //
+    // The edges belonging to RRNodeId is [first_edge, last_edge), excluding
+    // last_edge.
+    //
+    // If first_edge == last_edge, then a RRNodeId has no edges.
+    RREdgeId first_edge(const RRNodeId& id) const {
+        return first_edge_[id];
+    }
+    RREdgeId last_edge(const RRNodeId& id) const {
```

Comment:
Return the first_edge of the next rr_node, which is one past the edge id range for the node we care about. This implies we have one dummy rr_node at the end of first_edge_, and we always allocate that dummy node. We also assume that the edges have been sorted by rr_node, which is true after partition_edges().

Reply:
done